src/hotspot/share/code/compressedStream.hpp

```diff
@@ -111,4 +112,102 @@ class CompressedWriteStream : public CompressedStream {
   }
 };
 
+class CompressedSparseData : public ResourceObj {
+protected:
+  u_char* _buffer;
+  int     _position; // current byte offset
+  size_t  _bit_position {0}; // current bit offset
+
+public:
+  CompressedSparseData(u_char* buffer = NULL, int position = 0) {
+    _buffer   = buffer;
+    _position = position;
+  }
+
+  u_char* buffer() const { return _buffer; }
+};
+
+// Modified compression algorithm for a data set in which a significant part of the data is null
+class CompressedSparseDataReadStream : public CompressedSparseData {
+public:
+  CompressedSparseDataReadStream(u_char* buffer, int position) : CompressedSparseData(buffer, position) {}
+
+  void set_position(int pos) {
+    _bit_position = 0;
+    _position = pos;
+  }
+
+  jboolean read_bool()       { return read_int(); }
+  jbyte    read_byte()       { return read_int(); }
+  jint     read_signed_int() { return UNSIGNED5::decode_sign(read_int()); }
+  jint     read_int();
+  jdouble  read_double() {
+    jint h = reverse_bits(read_int());
+    jint l = reverse_bits(read_int());
+    return jdouble_cast(jlong_from(h, l));
+  }
+  jlong    read_long() {
+    jint low  = read_signed_int();
+    jint high = read_signed_int();
+    return jlong_from(high, low);
+  }
+
+protected:
+  bool read_zero();
+  uint8_t read_byte_impl();
+  inline u_char read()       { return _buffer[_position++]; }
+};
+
+class CompressedSparseDataWriteStream : public CompressedSparseData {
+public:
+  CompressedSparseDataWriteStream(int initial_size) : CompressedSparseData() {
+    _buffer   = NEW_RESOURCE_ARRAY(u_char, initial_size);
+    _size     = initial_size;
+    _buffer[0] = 0;
+  }
+
+  void write_bool(jboolean value)   { write_int(value ? 1 : 0); }
+  void write_byte(jbyte value)      { write_int(value); }
+  void write_signed_int(jint value) { write_int(UNSIGNED5::encode_sign(value)); }
+  void write_int(juint value);
+  void write_double(jdouble value)  {
+    juint rh = reverse_bits(high(jlong_cast(value)));
+    juint rl = reverse_bits(low( jlong_cast(value)));
+    write_int(rh);
+    write_int(rl);
+  }
+  void write_long(jlong value)      {
+    write_signed_int(low(value));
+    write_signed_int(high(value));
+  }
+
+  void align() {
+    if (_bit_position > 0) {
+      // start a new byte
+      _position++;
+      _bit_position = 0;
+    }
+  }
+  int position() {
+    align(); // method positon() have a side effect: the current data becomes aligned
+    return _position;
+  }
+  void set_position(int pos) {
```

Comment:
I have an idea which might solve the issues.

Reply:
> // Roll the stream state back to the marked one.  
> void roll_back();

get_position() is not about roll back only. See the DebugInformationRecorder, it serializes offsets into the stream.

```
int DebugInformationRecorder::serialize_scope_values(...) {
  ...
  int result = stream()->position();
  ...
  return result;
}

DebugToken* DebugInformationRecorder::create_scope_values(...) {
  ...
  return (DebugToken*) (intptr_t) serialize_scope_values(values);
}

void PhaseOutput::Process_OopMap_Node(MachNode *mach, int current_offset) {
  ...
  DebugToken *locvals = C->debug_info()->create_scope_values(locarray);
  DebugToken *expvals = C->debug_info()->create_scope_values(exparray);
  DebugToken *monvals = C->debug_info()->create_monitor_values(monarray);

  C->debug_info()->describe_scope(
    ...
    locvals,
    expvals,
    monvals
  );
  
void DebugInformationRecorder::describe_scope(...
                                              DebugToken* locals,
                                              DebugToken* expressions,
                                              DebugToken* monitors) {
  ...
  // serialize the locals/expressions/monitors
  stream()->write_int((intptr_t) locals);
  stream()->write_int((intptr_t) expressions);
  stream()->write_int((intptr_t) monitors);
```

Reply:
Thank you for the information. It's very helpful.

I think we should not simulate `CompressedWriteStream`. 

`DebugInformationRecorder` needs certain operations:
We write debug info into a stream writer: as grouped multiple data and single data. We need to know where bytes of grouped data begin and end. We need to keep offsets of grouped data in the stream. We need to be able to discard last written grouped data. We need to get the number of used bytes. We don't need to know how data stored in a stream.

Based on the specification, we need a stream writer to provide operations:
```
// Start grouped data.
// Return a position (byte offset) in the stream where grouped data begins.
int start_group();

// Finish grouped data.
// Return a position (byte offset) in the stream where grouped data ends.
int finish_group();

// Revert the stream to the specified position.
void set_position(int pos);

// Return the number of bytes stored data uses.
int data_size() const;
```

With them we don't have a function which in one implementation is const but in another implementation is with side effects. IMHO, at some point later side effects will cause bugs. 

Possible implementations:
```
int start_group() {
  complete_current_byte(); // this is renamed align()
  return _position;
}

int finish_group() {
  complete_current_byte();
  return _position;
}

int data_size() const {
  if (_position == 0 && _bit_position == 0)
    return 0;

 int used_bytes = _position;
 if (_bit_position != 0)
   ++used_bytes;
 return used_bytes;
}
```

In `DebugInformationRecorder` we will need to replace `position()` with `start_group()` and add `finish_group()`.
We will need to change `int DebugInformationRecorder::find_sharable_decode_offset(int stream_offset)` to
`int DebugInformationRecorder::find_sharable_decode_offset(int data_begin_offset, int data_end_offset)`.

If we want `DebugInformationRecorder` to use `CompressedWriteStream` we can use an adapter:
```
class CompressedWriteStreamAdapter: public CompressedWriteStream {
public:
...
int start_group() {
  return position();
}

int finish_group() {
  return position();
}

int data_size() const {
  return position();
}
 
};
```

Reply:
Sorry for not being online for a while :)
As you proposed, I refactored the CompressedSparseDataWriteStream interface.
I removed position() and set_position(int) methods. I added start_scope() and roll_back(int) instead.
Plus I have added is_empty() method (without side effect) to avoid unintentional align from assert.
Plus I added data_size() method (without side effect) - it is used to check the amount of data to copy.
Please check if it is Ok now.
Thanks.